kube-core/src/discovery.rs

```diff
@@ -55,6 +58,67 @@ impl ApiResource {
     pub fn from_gvk(gvk: &GroupVersionKind) -> Self {
         ApiResource::from_gvk_with_plural(gvk, &to_plural(&gvk.kind.to_ascii_lowercase()))
     }
+
+    /// Returns [`ApiResource`] instance for the custom resource,
+    /// defined by the CRD. You may use this `ApiResource` to manipulate
+    /// instances of the CRD.
+    /// ```no_run
+    /// # use kube::{Api, Client, core::{ApiResource, DynamicObject, ResourceExt, discovery::ApiResourceFromCrdHint}};
+    /// # use k8s_openapi::apiextensions_apiserver::pkg::apis::apiextensions::v1::CustomResourceDefinition;
+    /// # use std::error::Error;
+    /// async fn print_instances(k: &kube::Client, crd: &CustomResourceDefinition) -> Result<(), Box<dyn Error>> {
+    ///     let api_resource = ApiResource::from_crd(crd, &ApiResourceFromCrdHint::SingleVersioned)
+    ///         .expect("in this example we assume crd has one version");
+    ///     let api = Api::<DynamicObject>::all_with(k.clone(), &api_resource);
+    ///     for item in api.list(&Default::default()).await? {
+    ///         println!("{}", item.name_unchecked());
+    ///     }
+    ///     Ok(())
+    /// }
+    /// ```
+    /// # Hint
+    /// If crd has multiple versions, than it defines several api resources in fact.
+    /// Therefore this function takes additional `hint` parameter. See [`ApiResourceFromCrdHint`]
+    /// for effect of each variant.
+    pub fn from_crd(crd: &CustomResourceDefinition, hint: &ApiResourceFromCrdHint) -> Option<ApiResource> {
+        let kind = crd.spec.names.kind.clone();
+        let group = crd.spec.group.clone();
+        let plural = crd.spec.names.plural.clone();
+        let version = hint.select(crd)?.name.clone();
+        let gvk = GroupVersionKind { group, version, kind };
+
+        Some(ApiResource::from_gvk_with_plural(&gvk, &plural))
+    }
+}
+
+/// Hint for [`ApiResource::from_crd`]
+pub enum ApiResourceFromCrdHint {
+    /// Assume that CRD has only one version.
+    /// If this assumption is violated, None will be returned.
+    SingleVersioned,
+    /// Use given version.
+    /// If CRD does not have this version, None will be returned.
+    Exact(String),
+    /// Use storage version. This hint always succeeds for valid CRDs.
+    Storage,
+}
+
```

Comment:
I think we can drop this public type in favour of an `version: Option<&str>` and if it's none use storage version.
Kubernetes already ensures there's only one stored version, so I don't see the distinction between `SingleVersioned` and `Storage`.

Reply:
Done.